contrail-kubernetes-master/hooks/contrail_kubernetes_master_utils.py

```diff
@@ -0,0 +1,162 @@
+import json
+from socket import inet_aton
+import struct
+import time
+import os
+import base64
+
+from charmhelpers.core.hookenv import (
+    Hooks,
+    UnregisteredHookError,
+    config,
+    log,
+    relation_get,
+    related_units,
+    relation_ids,
+    status_set,
+    relation_set,
+    is_leader,
+    leader_get,
+    leader_set,
+    unit_private_ip,
+    charm_dir,
+
+)
+from charmhelpers.core.templating import render
+
+import common_utils
+import docker_utils
+from subprocess import (
+    check_output,
+)
+
+
+config = config()
+
+
+BASE_CONFIGS_PATH = "/etc/contrail"
+
+CONFIGS_PATH = BASE_CONFIGS_PATH + "/contrail-kubernetes"
+IMAGES = [
+        "contrail-kubernetes-kube-manager",
+    ]
+
+
+def kubernetes_token():
+    try:
+        account_file = os.path.join(charm_dir(), 'files', 'contrail-kubemanager-serviceaccount.yaml')
+        check_output(["snap", "run", "kubectl", "apply", "-f", account_file])
+    except:
+        return None
+    token_id = None
+    for i in range (10):
+        try:
+            token_id = check_output(["snap", "run", "kubectl", "get", "sa", "contrail-kubemanager", "-n", "contrail",
+                                "-ogo-template=\"{{(index .secrets 0).name}}\""]).strip('\"')
+        except:
+            return None
+        if token_id:
+            break
+        time.sleep(1)
+    if not token_id:
+        return None
+    try:
+        token_64 = check_output(["snap", "run", "kubectl", "get", "secret", token_id, "-n", "contrail",
+                            "-ogo-template=\"{{.data.token}}\""]).strip('\"')
+        token = base64.b64decode(token_64)
+        return token
+    except:
+        return None
+
+
+def get_context():
+    ctx = {}
+    ctx["log_level"] = config.get("log-level", "SYS_NOTICE")
+    ctx["container_registry"] = config.get("docker-registry")
+    ctx["contrail_version_tag"] = config.get("image-tag")
+
+    ctx["cluster_name"] = config.get("cluster_name")
+    ctx["cluster_project"] = config.get("cluster_project")
+    ctx["pod_subnets"] = config.get("pod_subnets")
+    ctx["ip_fabric_subnets"] = config.get("ip_fabric_subnets")
+    ctx["service_subnets"] = config.get("service_subnets")
+    ctx["ip_fabric_forwarding"] = config.get("ip_fabric_forwarding")
+    ctx["ip_fabric_snat"] = config.get("ip_fabric_snat")
+    ctx["host_network_service"] = config.get("host_network_service")
+    ctx["public_fip_pool"] = config.get("public_fip_pool")
+
+    ctx.update(common_utils.json_loads(leader_get("orchestrator-info"), dict()))
+
+    ips = [relation_get("private-address", unit, rid)
+           for rid in relation_ids("contrail-controller")
+           for unit in related_units(rid)]
+    ctx["controller_servers"] = ips
+    for rid in relation_ids("contrail-controller"):
```

Comment:
it should be in hook... like for vrouter-agent

Reply:
https://github.com/Juniper/contrail-charms/blob/R5/contrail-agent/hooks/contrail_agent_utils.py#L134-L135